File: WINGS_AI_Bootcamp/Elixir/roundTo5.exs
defmodule Rounding do
  def round_to_5(amt) do
    change = rem(round(amt * 100), 500) / 100.0
    fives = div(round(amt * 100), 500)
    if change >= 2.5, do: 5 * (fives + 1), else: 5 * fives
  end

  def nearest(amt, denom) do
    threshold = denom / 2
    how_many = floor(amt / denom)
    remainder = rem(round(amt * 100), denom * 100) / 100.0
    if remainder < threshold, do: how_many * denom, else: (how_many + 1) * denom
  end
end

# Hardcoded test values
amt1 = 42.73
amt2 = 42.73
amt3 = 42.73

IO.puts(Rounding.round_to_5(amt1))
IO.puts(Rounding.round_to_5(Rounding.nearest(amt2, 5)))
IO.puts(Rounding.round_to_5(Rounding.nearest(amt3, 10)))

File: WINGS_AI_Bootcamp/Haskell/roundTo5.hs
import Data.Fixed (mod')

roundTo5 :: Double -> Int
roundTo5 amt =
  let change = amt `mod'` 5
      fives = amt / 5
      threshold = 2.5
  in if change >= threshold
     then 5 * (floor fives + 1)
     else 5 * floor fives

nearest :: Double -> Int -> Int
nearest amt denom =
  let threshold = fromIntegral denom / 2
      howMany = floor (amt / fromIntegral denom)
  in if amt `mod'` fromIntegral denom < threshold
     then howMany * denom
     else (howMany + 1) * denom

nearestTo :: Double -> Int -> Int
nearestTo amt denom = round (amt / fromIntegral denom) * denom

main :: IO ()
main = do
  let amt1 = 42.73
  print (roundTo5 amt1)

  let amt2 = 42.73
      denom1 = 5
  print (roundTo5 (fromIntegral (nearest amt2 denom1)))

  let amt3 = 42.73
      denom2 = 10
  print (roundTo5 (fromIntegral (nearest amt3 denom2)))

